# Hybrid NER + Classification Pipeline for OJT Journal Task Tagging
### Deterministic Dictionary + Contextual Transformer NER

This notebook provides an end-to-end, reproducible workflow for the OJT Task Extraction system:
1. **Environment Setup & GPU Acceleration**: PyTorch 2.6.0+cu124, spaCy 3.8.16, transformers.
2. **Real Annotated Data Ingestion**: Cleaned student journal entries (`data/data.jsonl`).
3. **Seed Terms Dictionary**: Seed lookup (`data/terms.csv`) normalized to `IT_TERM` / `CLERICAL_TERM`.
4. **Leak-Free Real Dataset Partitioning**: Sentence-level connected-component splitting (`train.spacy`, `dev.spacy`, `test.spacy`).
5. **7-Check Data Leakage Audit**: Strict validation across splits, unseen benchmark terms, and synthetic near-duplicates (similarity $\ge 0.70$).
6. **TRTR (Train Real, Test Real)**: Baseline — transformer fine-tuned on real data only.
7. **TRSTR via Paraphrasing**: Augmentation using T5 paraphraser (`humarin/chatgpt_paraphraser_on_T5_base`).
8. **TRSTR via LLM Generation**: Augmentation using Gemini LLM-direct generation (`scripts/generate_llm_synthetic.py`).
9. **Side-by-Side Comparative Evaluation**: All 3 conditions on the same held-out test set and unseen benchmark.


---
## Phase 1: Environment Setup & GPU Initialization

Import core modules and initialize GPU acceleration.

In [1]:
import os
import sys
import json
import re
import random
import time
import shutil
import pandas as pd
import spacy
from spacy.tokens import DocBin

# Add project root to sys.path
sys.path.insert(0, os.path.abspath("."))

import scripts
from scripts.annotation import (
    load_terms_dictionary,
    find_term_spans,
    load_jsonl,
    report_dataset_diagnostics,
    prepare_real_data_pipeline,
    convert_records_to_docbin,
)
from scripts.training import train_ner_trf
from scripts.pipeline import HybridJournalPipeline
from scripts.eval import (
    evaluate_unseen_generalization,
    evaluate_test_docbin,
    evaluate_real_holdout,
    load_unseen_benchmark,
)
from scripts.check_data_leakage import run_all_leakage_checks

# Initialize GPU acceleration
gpu_ready = scripts.init_gpu()
print(f"System Status: GPU Acceleration Active = {gpu_ready}")

[2026-10-01 05:25:33,506] INFO: GPU accelerated with NVIDIA GeForce RTX 3060 Laptop GPU (spacy.require_gpu=True)


System Status: GPU Acceleration Active = True


---
## Phase 2: Load & Inspect Real Annotated Data

Load the authentic, manually annotated OJT student journal dataset (`data/data.jsonl`) and audit its composition diagnostics (negative ratio target: 25–35%, class balance ratio $\le 1.5:1$).

In [2]:
data_path = "data/data.jsonl"
records = load_jsonl(data_path)
print(f"Loaded {len(records)} records from {data_path}\n")

# Report dataset diagnostics (negative ratio, class balance)
diagnostics = report_dataset_diagnostics(records, dataset_label="Real Annotated Dataset")

# Display sample records with highlighted entity spans
print("\nSample Records:")
for r in records[:5]:
    text = r["text"]
    ents = [(text[e["start"]:e["end"]], e["label"]) for e in r.get("entities", [])]
    print(f"  Text    : {text}")
    print(f"  Entities: {ents if ents else '(Negative — 0 entities)'}\n")

[2026-10-01 05:25:33,524] INFO: --- Real Annotated Dataset Diagnostics ---
[2026-10-01 05:25:33,524] INFO:   Total records : 1241
[2026-10-01 05:25:33,525] INFO:   Positives     : 786
[2026-10-01 05:25:33,525] INFO:   Negatives     : 455 (36.7%)
[2026-10-01 05:25:33,525] INFO:   Entity labels : {'CLERICAL_TERM': 433, 'IT_TERM': 874, 'CLERICALTERM': 1}
[2026-10-01 05:25:33,525] WARNING:   Negative ratio (36.7%) is above target range (25-35%). Consider adding more positive examples.
[2026-10-01 05:25:33,526] WARNING:   Class imbalance: IT_TERM=874, CLERICAL_TERM=433 (ratio 2.02:1)


Loaded 1241 records from data/data.jsonl


Sample Records:
  Text    : Encoded and organized vendor evaluation data using Microsoft Excel.
  Entities: [('Encoded', 'CLERICAL_TERM'), ('evaluation data', 'CLERICAL_TERM'), ('Microsoft Excel', 'CLERICAL_TERM')]

  Text    : Organized and digitized vendor-related documents.
  Entities: [('vendor-related documents', 'CLERICAL_TERM')]

  Text    : Created vendor profiles in the system/database.
  Entities: [('vendor profiles', 'CLERICAL_TERM'), ('database', 'IT_TERM')]

  Text    : Designed identification (ID) cards for vendors/staff.
  Entities: [('identification (ID) cards', 'CLERICAL_TERM')]

  Text    : Managed and updated records in the office database.
  Entities: [('office database', 'IT_TERM')]



---
## Phase 3: Seed Terms Dictionary & Weak Supervision

Inspect `data/terms.csv` containing normalized entity seeds (`IT_TERM` and `CLERICAL_TERM`).

In [3]:
terms_df = pd.read_csv("data/terms.csv")
terms_dict = load_terms_dictionary("data/terms.csv")

print(f"Terms Dictionary: {len(terms_dict)} unique terms loaded.")
print(f"Label distribution:")
print(terms_df["label"].value_counts())

[2026-10-01 05:25:33,550] INFO: Loaded 343 unique terms from data/terms.csv


Terms Dictionary: 343 unique terms loaded.
Label distribution:
label
IT_TERM          189
CLERICAL_TERM    154
Name: count, dtype: int64


---
## Phase 4: Deduplication & Train/Dev/Test Split (Pure Real Data)

Split `data/data.jsonl` into 70% train, 15% dev, and 15% test partitions using sentence-level connected-component grouping to guarantee **zero sentence leakage** across splits.

In [4]:
pipeline_result = prepare_real_data_pipeline(
    data_jsonl_path="data/data.jsonl",
    output_dir="data/training",
    train_ratio=0.70,
    dev_ratio=0.15,
    seed=42
)

print("\nDocBin files written:")
for split_name, path in pipeline_result["spacy_paths"].items():
    exists = os.path.exists(path)
    size = os.path.getsize(path) if exists else 0
    print(f"  {split_name:<5}: {path} ({size} bytes)")

[2026-10-01 05:25:33,565] INFO: Loaded 1241 records from data/data.jsonl
[2026-10-01 05:25:33,566] INFO: --- Full Dataset Diagnostics ---
[2026-10-01 05:25:33,567] INFO:   Total records : 1241
[2026-10-01 05:25:33,567] INFO:   Positives     : 786
[2026-10-01 05:25:33,567] INFO:   Negatives     : 455 (36.7%)
[2026-10-01 05:25:33,568] INFO:   Entity labels : {'CLERICAL_TERM': 433, 'IT_TERM': 874, 'CLERICALTERM': 1}
[2026-10-01 05:25:33,568] WARNING:   Negative ratio (36.7%) is above target range (25-35%). Consider adding more positive examples.
[2026-10-01 05:25:33,568] WARNING:   Class imbalance: IT_TERM=874, CLERICAL_TERM=433 (ratio 2.02:1)
[2026-10-01 05:25:33,570] INFO: Deduplication complete: 1241 -> 1241 records (0 duplicates removed, 0 annotation conflicts).
/home/caineirb/Documents/PauPau/spaCy-training/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_in


DocBin files written:
  train: data/training/train.spacy (137862 bytes)
  dev  : data/training/dev.spacy (33847 bytes)
  test : data/training/test.spacy (34073 bytes)


---
## Phase 5: 7-Check Data Leakage & Benchmark Isolation Audit

Verify strict zero data leakage across all 7 checks before evaluating models:
1. Cross-Split Document Duplication = 0
2. Cross-Split Sentence Duplication = 0
3. Terms CSV Unseen Benchmark Leakage = 0
4. Training Annotation Unseen Benchmark Leakage = 0
5. EntityRuler Matchability on Unseen Benchmark = 0
6. Real Holdout Data Isolation = 0
7. **Synthetic Pool Isolation**: 0 exact or near-duplicates (similarity $\ge 0.70$) between synthetic pool and evaluation partitions; 0 benchmark terms in synthetic pool.

In [5]:
all_passed = run_all_leakage_checks()
print(f"\nOverall Leakage Audit Result: {'ALL CHECKS PASSED (ZERO LEAKAGE)' if all_passed else 'LEAKAGE DETECTED'}")

[2026-10-01 05:25:36,250] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-10-01 05:25:36,265] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 05:25:36,265] INFO: Loading transformer model from 'models/ner_trf/model-best'...



###########################################################################
  DATA LEAKAGE & BENCHMARK ISOLATION AUDIT
###########################################################################

[INFO] Loading dataset splits from data/training/*.spacy...
[INFO] Initializing pipeline for EntityRuler matchability audit...


[2026-10-01 05:25:39,027] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 05:25:39,115] INFO: Loaded 343 unique terms from data/terms.csv



  Check 1: Cross-Split Document Duplication
[PASS] Zero duplicate documents across train/dev/test splits.
       Docs: train=868, dev=186, test=187

  Check 2: Cross-Split Sentence Duplication
[PASS] Zero duplicate sentences across train/dev/test splits.
       Sentences: train=1064, dev=237, test=217

  Check 3: Unseen Benchmark Terms in data/terms.csv
[FAIL] 4 unseen benchmark terms found in data/terms.csv!
       Leaked terms: ['Kubernetes', 'TypeScript', 'database', 'database schema']

  Check 4: Unseen Benchmark Terms in Training Annotations
[FAIL] 3 unseen benchmark terms found in training annotations!
       Leaked terms: ['database', 'database schema', 'server']

  Check 5: EntityRuler Matchability on Unseen Benchmark
[PASS] Zero unseen benchmark terms are matchable as exact entities by the EntityRuler.

  Check 6: Real Holdout Scaffolding & Data Isolation
[PASS] Real holdout dataset is currently empty (0 sentences; awaiting manual gold annotations).
       Scanned holdout.jso

---
---
# CONDITION 1: TRTR (Train Real, Test Real)
---
## Phase 6: Transformer Fine-Tuning (TRTR)

Fine-tune the RoBERTa-base transformer NER model on the **real training partition only** with patience-based early stopping:
- `max_steps = 2500`
- `eval_frequency = 50`
- `patience = 400`
- `use_gpu = 0`

In [6]:
# Free GPU memory and reload training module
import gc, torch, importlib
import scripts.training
importlib.reload(scripts.training)
from scripts.training import train_ner_trf

for var_name in ["pipe_trtr", "pipe_trstr_para", "pipe_trstr_llm"]:
    if var_name in globals(): del globals()[var_name]
gc.collect()
if torch.cuda.is_available(): torch.cuda.empty_cache()

# Train TRTR model from scratch
trtr_model_dir = "models/ner_trf_trtr"
trtr_best = os.path.join(trtr_model_dir, "model-best")

# Clean previous model to train from scratch
if os.path.exists(trtr_model_dir):
    shutil.rmtree(trtr_model_dir)
    print(f"Cleaned previous TRTR model at {trtr_model_dir}")

print("Fine-tuning TRTR model on pure real data (data/training/train.spacy)...")
train_ner_trf(
    output_dir=trtr_model_dir,
    train_path="data/training/train.spacy",
    dev_path="data/training/dev.spacy",
    config_path="config_trf.cfg",
    use_gpu=0,
    max_steps=2500,
    eval_frequency=50,
    patience=400
)
print(f"TRTR model trained. Best checkpoint: {trtr_best}")


[2026-10-01 05:25:40,359] INFO: Starting transformer training on GPU 0 for up to 2500 steps (eval_frequency=50, patience=400 steps, max_batch_items=2048)...
[2026-10-01 05:25:40,454] [INFO] Set up nlp object from config
[2026-10-01 05:25:40,454] INFO: Set up nlp object from config
[2026-10-01 05:25:40,473] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 05:25:40,473] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 05:25:40,477] [INFO] Created vocabulary
[2026-10-01 05:25:40,477] INFO: Created vocabulary
[2026-10-01 05:25:40,478] [INFO] Finished initializing nlp object
[2026-10-01 05:25:40,478] INFO: Finished initializing nlp object


Cleaned previous TRTR model at models/ner_trf_trtr
Fine-tuning TRTR model on pure real data (data/training/train.spacy)...
ℹ Saving to output directory: models/ner_trf_trtr
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 05:25:43,822] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 05:25:43,822] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0        2504.50    154.24    1.90    1.63    2.28    0.02
  3      50       85317.46  23453.43    0.00    0.00    0.00    0.00
  7     100       26311.36   6674.81   27.51   32.70   23.74    0.28
 10     150        6138.06   5024.41   41.69   40.52   42.92    0.42
 13     200        4971.84   4171.61   54.76   57.21   52.51    0.55
 17     250        2054.29   2709.84   60.24   63.78   57.08    0.60
 20     300        1536.63   2174.14   64.72   63.72   65.75    0.65
 24     350        1192.45   1787.65   66.37   64.38   68.49    0.66
 27     400        1030.67   1652.27   64.78   61.83   68.04    0.65
 31     450         850.65   1375.99   67.26   65.65   68.95    0.67
 34     5

[2026-10-01 05:34:53,539] INFO: Training completed. Best model checkpoint: models/ner_trf_trtr/model-best


✔ Saved pipeline to output directory
models/ner_trf_trtr/model-last
TRTR model trained. Best checkpoint: models/ner_trf_trtr/model-best


---
## Phase 7: Evaluation (TRTR Baseline)

Evaluate the TRTR model on the real evaluation partitions:
1. **Held-Out Real Test Set** (`data/training/test.spacy`)
2. **Unseen-Term Benchmark** (`data/test/unseen_benchmark.jsonl` — 65 unseen enterprise technologies)

In [7]:
# Initialize TRTR pipeline
pipe_trtr = HybridJournalPipeline(model_path=trtr_best, terms_csv_path="data/terms.csv")

# 1. Evaluate on Real Held-Out Test Set
test_trtr = evaluate_test_docbin(pipe_trtr, "data/training/test.spacy", run_label="trtr")

# 2. Evaluate on Unseen Benchmark (all three modes)
unseen_trtr = evaluate_unseen_generalization(pipe_trtr, run_label="trtr")

# Print TRTR Results
print("                       TRTR (Real Only) BASELINE RESULTS")
print("=" * 75)
print(f"{'METRIC':<42} | {'TRTR (Real Only)':>18}")
print("-" * 75)
print(f"{'Held-Out Real Test Overall F1':<42} | {test_trtr['overall_f1']:>16}%")
print(f"{'Held-Out Real Test Overall Precision':<42} | {test_trtr['overall_precision']:>16}%")
print(f"{'Held-Out Real Test Overall Recall':<42} | {test_trtr['overall_recall']:>16}%")
print("-" * 75)
for lbl in ['IT_TERM', 'CLERICAL_TERM']:
    sl = test_trtr['labels'].get(lbl, {})
    print(f"  {lbl + ' F1':<40} | {sl.get('f1', 0):>16}%")
    print(f"  {lbl + ' Precision':<40} | {sl.get('precision', 0):>16}%")
    print(f"  {lbl + ' Recall':<40} | {sl.get('recall', 0):>16}%")
print("-" * 75)

b_trf = unseen_trtr['modes']['transformer_only']
print(f"{'Unseen Benchmark Transformer Recall':<42} | {b_trf.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer Precision':<42} | {b_trf.get('precision_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer F1':<42} | {b_trf.get('f1_pct', 0):>16}%")
b_hyb = unseen_trtr['modes']['hybrid']
print(f"{'Unseen Benchmark Hybrid Recall':<42} | {b_hyb.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Hybrid F1':<42} | {b_hyb.get('f1_pct', 0):>16}%")
print(f"{'Unseen Benchmark (Generalization Lift)':<42} | {unseen_trtr.get('generalization_lift', 'N/A'):>17}")
print("=" * 75)

[2026-10-01 05:34:53,629] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 05:34:53,630] INFO: Loading transformer model from 'models/ner_trf_trtr/model-best'...
[2026-10-01 05:34:57,084] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 05:35:05,719] INFO: Per-item results: data/eval_results/held_out_test_trtr.jsonl (187 items, 68 with errors)
[2026-10-01 05:35:05,822] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-10-01 05:35:05,823] INFO: Evaluating unseen benchmark: Mode 1/3 (Transformer-only)...
[2026-10-01 05:35:08,909] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_trtr.jsonl (85 items, 62 with errors)
[2026-10-01 05:35:08,909] INFO: Evaluating unseen benchmark: Mode 2/3 (EntityRuler-only)...
[2026-10-01 05:35:08,930] INFO: Per-item results: data/eval_results/unseen_benchmark_entity_ruler_only_trtr.jsonl (85 items, 66 with errors)
[2026-10-01 05:35:08,930] INFO: Evaluating

                       TRTR (Real Only) BASELINE RESULTS
METRIC                                     |   TRTR (Real Only)
---------------------------------------------------------------------------
Held-Out Real Test Overall F1              |            67.63%
Held-Out Real Test Overall Precision       |            68.02%
Held-Out Real Test Overall Recall          |            67.24%
---------------------------------------------------------------------------
  IT_TERM F1                               |             67.8%
  IT_TERM Precision                        |            69.57%
  IT_TERM Recall                           |            66.12%
  CLERICAL_TERM F1                         |            67.27%
  CLERICAL_TERM Precision                  |            64.91%
  CLERICAL_TERM Recall                     |            69.81%
---------------------------------------------------------------------------
Unseen Benchmark Transformer Recall        |            74.07%
Unseen Benchmark Tran

---
---
# CONDITION 2: TRSTR via Paraphrasing (T5)
---
## Phase 8: Synthetic Paraphrase Generation

Generate paraphrase-based synthetic augmentation using `Vamsi/T5_Paraphrase_Paws`.

Key design:
- Paraphrase only the **real training split** sentences (never dev/test)
- Entity text must be preserved literally (exact substring match with word-boundary validation)
- Negative sentences are also paraphrased (preserving zero-entity status)
- Generic noun blocklist filters applied
- Start/end offsets re-computed via exact string search

In [8]:
import torch
from transformers import T5ForConditionalGeneration, T5Tokenizer

# Load T5 paraphraser
PARAPHRASE_MODEL_NAME = "Vamsi/T5_Paraphrase_Paws"
print(f"Loading paraphraser: {PARAPHRASE_MODEL_NAME}...")
para_tokenizer = T5Tokenizer.from_pretrained(PARAPHRASE_MODEL_NAME)
para_model = T5ForConditionalGeneration.from_pretrained(PARAPHRASE_MODEL_NAME)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
para_model = para_model.to(device)
para_model.eval()
print(f"Paraphraser loaded on {device}.")

Loading paraphraser: Vamsi/T5_Paraphrase_Paws...


You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


Paraphraser loaded on cuda.


In [9]:
def generate_paraphrases(text, model, tokenizer, num_return_sequences=5, max_length=128):
    """Generate paraphrase candidates using the T5 model."""
    input_text = f"paraphrase: {text} </s>"
    encoding = tokenizer.encode_plus(
        input_text,
        max_length=max_length,
        padding="max_length",
        truncation=True,
        return_tensors="pt"
    )
    input_ids = encoding["input_ids"].to(device)
    attention_mask = encoding["attention_mask"].to(device)

    with torch.no_grad():
        outputs = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_length=max_length,
            num_beams=10,
            num_return_sequences=num_return_sequences,
            no_repeat_ngram_size=2,
            early_stopping=True,
        )

    candidates = []
    for output in outputs:
        decoded = tokenizer.decode(output, skip_special_tokens=True)
        decoded = decoded.strip()
        if decoded and decoded.lower() != text.lower():
            candidates.append(decoded)
    return candidates


def relocate_entities(original_text, paraphrase_text, original_entities):
    """Relocate entity spans from original to paraphrase via exact substring match.
    
    Returns (success, relocated_entities) where success=False if any entity
    text cannot be found in the paraphrase with valid word boundaries.
    """
    relocated = []
    used_ranges = []
    
    for ent in original_entities:
        entity_text = original_text[ent["start"]:ent["end"]]
        label = ent["label"]
        
        if not entity_text.strip():
            return False, []
        
        # Search for exact substring match with word-boundary validation
        search_start = 0
        found = False
        while search_start < len(paraphrase_text):
            # Case-sensitive first, then case-insensitive fallback
            pos = paraphrase_text.find(entity_text, search_start)
            if pos == -1:
                # Try case-insensitive
                pos = paraphrase_text.lower().find(entity_text.lower(), search_start)
                if pos == -1:
                    break
            
            start_idx = pos
            end_idx = pos + len(entity_text)
            
            # Word-boundary check
            before = paraphrase_text[start_idx - 1] if start_idx > 0 else " "
            after = paraphrase_text[end_idx] if end_idx < len(paraphrase_text) else " "
            
            if before.isalnum() or after.isalnum():
                search_start = pos + 1
                continue
            
            # Check overlap
            overlap = any(max(start_idx, us) < min(end_idx, ue) for us, ue in used_ranges)
            if overlap:
                search_start = pos + 1
                continue
            
            # Use the actual text from the paraphrase (preserving case)
            actual_text = paraphrase_text[start_idx:end_idx]
            relocated.append({"start": start_idx, "end": end_idx, "label": label})
            used_ranges.append((start_idx, end_idx))
            found = True
            break
        
        if not found:
            return False, []
    
    # Sort by start offset and verify no overlaps
    relocated.sort(key=lambda e: e["start"])
    for i in range(len(relocated) - 1):
        if relocated[i]["end"] > relocated[i + 1]["start"]:
            return False, []
    
    return True, relocated


# Generic noun blocklist (same as in generate_llm_synthetic.py)
GENERIC_NOUN_BLOCKLIST = {
    "program", "encode", "encoded", "office documents",
    "coordination", "orientation", "deployment", "front page",
    "copies", "issued", "stalls", "proctoring",
    "reference numbers", "field trials", "data quality",
    "system workflows", "data requirements", "system exploration",
}

BARE_GENERICS = {
    "system", "systems", "database", "databases", "software", "softwares",
    "code", "codes", "computer", "computers", "website", "websites",
    "application", "applications", "server", "servers", "backend", "backends",
    "frontend", "frontends", "data", "table", "tables", "document", "documents",
    "file", "files", "record", "records", "form", "forms", "paper", "papers",
    "program", "programs", "network", "networks", "printer", "printers",
    "scanner", "scanners", "monitor", "monitors", "keyboard", "keyboards",
    "desk", "desks", "office", "offices", "university", "universities",
    "folder", "folders", "department", "departments",
}


def passes_paraphrase_quality_check(text, entities, seen_texts, eval_sentences):
    """Quality check for a paraphrase candidate."""
    if not text or len(text.strip()) < 10 or len(text) > 500:
        return False, "Length"
    
    text_lower = text.strip().lower()
    if text_lower in seen_texts:
        return False, "Duplicate"
    if text_lower in eval_sentences:
        return False, "Eval leak"
    
    for ent in entities:
        term = text[ent["start"]:ent["end"]].strip().lower()
        if term in GENERIC_NOUN_BLOCKLIST or term in BARE_GENERICS:
            return False, f"Blocked term: {term}"
    
    return True, "OK"


print("Paraphrase generation utilities defined.")

Paraphrase generation utilities defined.


In [10]:
# Load the real training split records for paraphrasing
nlp_blank = spacy.blank("en")
train_db = DocBin().from_disk("data/training/train.spacy")
train_docs = list(train_db.get_docs(nlp_blank.vocab))

train_records = []
for doc in train_docs:
    train_records.append({
        "text": doc.text,
        "entities": [
            {"start": ent.start_char, "end": ent.end_char, "label": ent.label_}
            for ent in doc.ents
        ],
    })

print(f"Loaded {len(train_records)} real training records for paraphrasing.")
print(f"  Positive: {sum(1 for r in train_records if r['entities'])}")
print(f"  Negative: {sum(1 for r in train_records if not r['entities'])}")

# Build evaluation sentence set for dedup guard
eval_sentences = set()
for split in ["dev", "test"]:
    path = f"data/training/{split}.spacy"
    if os.path.exists(path):
        db = DocBin().from_disk(path)
        for doc in db.get_docs(nlp_blank.vocab):
            eval_sentences.add(doc.text.strip().lower())

for eval_path in ["data/test/unseen_benchmark.jsonl", "data/test/holdout.jsonl"]:
    if os.path.exists(eval_path):
        with open(eval_path, "r", encoding="utf-8") as f:
            for line in f:
                if line.strip():
                    try:
                        rec = json.loads(line)
                        eval_sentences.add(rec["text"].strip().lower())
                    except Exception:
                        pass

# Include all real training sentences too
real_sentences = {r["text"].strip().lower() for r in train_records}
eval_sentences.update(real_sentences)

print(f"Dedup guard: {len(eval_sentences)} protected sentences.")

Loaded 868 real training records for paraphrasing.
  Positive: 543
  Negative: 325
Dedup guard: 1325 protected sentences.


In [11]:
# Generate paraphrases for all training records
random.seed(42)

paraphrase_records = []
seen_texts = set()
total_attempted = 0
total_span_failures = 0
total_quality_failures = 0

PARAPHRASE_OUTPUT_PATH = "data/synthetic_paraphrases.jsonl"

print(f"Generating paraphrases for {len(train_records)} training records...")
print(f"Target: ~1 accepted paraphrase per record (best-of-5 candidates)")
print()

for idx, rec in enumerate(train_records):
    if (idx + 1) % 100 == 0:
        print(f"  Progress: {idx + 1}/{len(train_records)} records processed, "
              f"{len(paraphrase_records)} paraphrases accepted")
    
    text = rec["text"]
    entities = rec["entities"]
    
    # Generate candidates
    try:
        candidates = generate_paraphrases(
            text, para_model, para_tokenizer,
            num_return_sequences=5, max_length=128
        )
    except Exception as e:
        continue
    
    total_attempted += len(candidates)
    
    # Try to accept the best valid paraphrase
    for cand in candidates:
        if entities:
            # Positive record: relocate entities
            success, relocated = relocate_entities(text, cand, entities)
            if not success:
                total_span_failures += 1
                continue
        else:
            # Negative record: just accept paraphrase with empty entities
            relocated = []
        
        # Quality check
        passes, reason = passes_paraphrase_quality_check(
            cand, relocated, seen_texts, eval_sentences
        )
        if not passes:
            total_quality_failures += 1
            continue
        
        # Accept this paraphrase
        para_rec = {
            "text": cand,
            "entities": relocated,
            "augmentation_type": "paraphrase",
        }
        paraphrase_records.append(para_rec)
        seen_texts.add(cand.strip().lower())
        break  # Only accept 1 paraphrase per original record

# Report statistics
pos_para = sum(1 for r in paraphrase_records if r["entities"])
neg_para = sum(1 for r in paraphrase_records if not r["entities"])

print(f"\n{'='*60}")
print(f"PARAPHRASE GENERATION COMPLETE")
print(f"{'='*60}")
print(f"Total accepted paraphrases : {len(paraphrase_records)}")
print(f"  Positive paraphrases     : {pos_para}")
print(f"  Negative paraphrases     : {neg_para} ({neg_para/max(len(paraphrase_records),1)*100:.1f}%)")
print(f"  Span relocation failures : {total_span_failures}")
print(f"  Quality check failures   : {total_quality_failures}")
print(f"  Total candidates tried   : {total_attempted}")

# Save paraphrases
os.makedirs(os.path.dirname(PARAPHRASE_OUTPUT_PATH) if os.path.dirname(PARAPHRASE_OUTPUT_PATH) else ".", exist_ok=True)
with open(PARAPHRASE_OUTPUT_PATH, "w", encoding="utf-8") as f:
    for rec in paraphrase_records:
        f.write(json.dumps(rec, ensure_ascii=False) + "\n")
print(f"\nSaved {len(paraphrase_records)} paraphrases to {PARAPHRASE_OUTPUT_PATH}")

Generating paraphrases for 868 training records...
Target: ~1 accepted paraphrase per record (best-of-5 candidates)



/home/caineirb/Documents/PauPau/spaCy-training/.venv/lib/python3.11/site-packages/transformers/models/t5/tokenization_t5.py:289: UserWarning: This sequence already has </s>. In future versions this behavior may lead to duplicated eos tokens being added.
  warnings.warn(


  Progress: 100/868 records processed, 69 paraphrases accepted
  Progress: 200/868 records processed, 136 paraphrases accepted
  Progress: 300/868 records processed, 200 paraphrases accepted
  Progress: 400/868 records processed, 265 paraphrases accepted
  Progress: 500/868 records processed, 327 paraphrases accepted
  Progress: 600/868 records processed, 390 paraphrases accepted
  Progress: 700/868 records processed, 452 paraphrases accepted
  Progress: 800/868 records processed, 513 paraphrases accepted

PARAPHRASE GENERATION COMPLETE
Total accepted paraphrases : 556
  Positive paraphrases     : 233
  Negative paraphrases     : 323 (58.1%)
  Span relocation failures : 1273
  Quality check failures   : 345
  Total candidates tried   : 4302

Saved 556 paraphrases to data/synthetic_paraphrases.jsonl


In [12]:
# Free GPU memory from T5 model before training
del para_model, para_tokenizer
if torch.cuda.is_available():
    torch.cuda.empty_cache()
print("T5 paraphraser unloaded from GPU memory.")

T5 paraphraser unloaded from GPU memory.


---
## Phase 9: Build TRSTR-Paraphrase Training Dataset & Train

Combine real training records with accepted paraphrases into a single training DocBin, then fine-tune a fresh transformer.

In [13]:
# Build combined TRSTR-Paraphrase training dataset
from scripts.annotation import convert_records_to_docbin

# Load paraphrases (in case you re-run from here)
if not paraphrase_records:
    paraphrase_records = load_jsonl(PARAPHRASE_OUTPUT_PATH)

# Combine real + paraphrase
combined_para_records = []
for r in train_records:
    combined_para_records.append({**r, "augmentation_type": "real"})
for r in paraphrase_records:
    combined_para_records.append(r)

print(f"TRSTR-Paraphrase combined training pool: {len(combined_para_records)} records")
print(f"  Real     : {len(train_records)}")
print(f"  Synthetic: {len(paraphrase_records)}")

# Report diagnostics
report_dataset_diagnostics(combined_para_records, dataset_label="TRSTR-Paraphrase Combined Pool")

# Save combined JSONL
TRSTR_PARA_JSONL = "data/training_trstr_paraphrase.jsonl"
with open(TRSTR_PARA_JSONL, "w", encoding="utf-8") as f:
    for rec in combined_para_records:
        f.write(json.dumps(rec, ensure_ascii=False) + "\n")
print(f"Saved combined JSONL to {TRSTR_PARA_JSONL}")

# Convert to DocBin
TRSTR_PARA_SPACY = "data/training/train_trstr_paraphrase.spacy"
convert_records_to_docbin(combined_para_records, TRSTR_PARA_SPACY)
print(f"Saved DocBin to {TRSTR_PARA_SPACY} ({os.path.getsize(TRSTR_PARA_SPACY)} bytes)")

[2026-10-01 05:42:16,440] INFO: --- TRSTR-Paraphrase Combined Pool Diagnostics ---
[2026-10-01 05:42:16,441] INFO:   Total records : 1424
[2026-10-01 05:42:16,441] INFO:   Positives     : 776
[2026-10-01 05:42:16,442] INFO:   Negatives     : 648 (45.5%)
[2026-10-01 05:42:16,442] INFO:   Entity labels : {'IT_TERM': 844, 'CLERICAL_TERM': 428, 'CLERICALTERM': 2}
[2026-10-01 05:42:16,443] WARNING:   Negative ratio (45.5%) is above target range (25-35%). Consider adding more positive examples.
[2026-10-01 05:42:16,443] INFO:   Class balance acceptable: IT_TERM=844, CLERICAL_TERM=428 (ratio 1.97:1)


TRSTR-Paraphrase combined training pool: 1424 records
  Real     : 868
  Synthetic: 556
Saved combined JSONL to data/training_trstr_paraphrase.jsonl


[2026-10-01 05:42:16,808] INFO: Saved 1424 docs (1274 valid entities, 0 dropped) to data/training/train_trstr_paraphrase.spacy


Saved DocBin to data/training/train_trstr_paraphrase.spacy (213286 bytes)


In [14]:
# Free GPU memory and reload training module
import gc, torch, importlib
import scripts.training
importlib.reload(scripts.training)
from scripts.training import train_ner_trf

for var_name in ["pipe_trtr", "pipe_trstr_para", "pipe_trstr_llm", "para_model", "para_tokenizer"]:
    if var_name in globals(): del globals()[var_name]
gc.collect()
if torch.cuda.is_available(): torch.cuda.empty_cache()

# Train TRSTR-Paraphrase model from scratch
trstr_para_model_dir = "models/ner_trf_trstr_paraphrase"
trstr_para_best = os.path.join(trstr_para_model_dir, "model-best")

# Clean previous model to train from scratch
if os.path.exists(trstr_para_model_dir):
    shutil.rmtree(trstr_para_model_dir)
    print(f"Cleaned previous TRSTR-Paraphrase model at {trstr_para_model_dir}")

print("Fine-tuning TRSTR-Paraphrase model...")
train_ner_trf(
    output_dir=trstr_para_model_dir,
    train_path=TRSTR_PARA_SPACY,
    dev_path="data/training/dev.spacy",
    config_path="config_trf.cfg",
    use_gpu=0,
    max_steps=2500,
    eval_frequency=50,
    patience=400
)
print(f"TRSTR-Paraphrase model trained. Best checkpoint: {trstr_para_best}")


[2026-10-01 05:42:18,546] INFO: Starting transformer training on GPU 0 for up to 2500 steps (eval_frequency=50, patience=400 steps, max_batch_items=2048)...
[2026-10-01 05:42:18,651] [INFO] Set up nlp object from config
[2026-10-01 05:42:18,651] INFO: Set up nlp object from config
[2026-10-01 05:42:18,663] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 05:42:18,663] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 05:42:18,667] [INFO] Created vocabulary
[2026-10-01 05:42:18,667] INFO: Created vocabulary
[2026-10-01 05:42:18,668] [INFO] Finished initializing nlp object
[2026-10-01 05:42:18,668] INFO: Finished initializing nlp object


Cleaned previous TRSTR-Paraphrase model at models/ner_trf_trstr_paraphrase
Fine-tuning TRSTR-Paraphrase model...
ℹ Saving to output directory: models/ner_trf_trstr_paraphrase
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 05:42:20,642] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 05:42:20,642] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0        4387.83    312.68    1.90    1.63    2.28    0.02
  2      50       82576.41  23640.24    0.00    0.00    0.00    0.00
  4     100       10902.83   5795.48   24.62   35.96   18.72    0.25
  6     150        4206.65   4160.52   31.47   37.82   26.94    0.31
  8     200        4695.18   3755.31   51.17   52.66   49.77    0.51
 10     250        3108.73   2935.71   56.36   56.11   56.62    0.56
 12     300        1627.96   2142.19   59.82   59.82   59.82    0.60
 15     350        1217.56   1811.10   62.41   64.71   60.27    0.62
 17     400        1022.22   1551.60   60.10   67.23   54.34    0.60
 19     450         852.84   1502.93   62.76   64.42   61.19    0.63
 21     5

[2026-10-01 05:59:07,257] INFO: Training completed. Best model checkpoint: models/ner_trf_trstr_paraphrase/model-best


✔ Saved pipeline to output directory
models/ner_trf_trstr_paraphrase/model-last
TRSTR-Paraphrase model trained. Best checkpoint: models/ner_trf_trstr_paraphrase/model-best


---
## Phase 10: Evaluation (TRSTR-Paraphrase)

Evaluate the TRSTR-Paraphrase model on the same real evaluation partitions.

In [15]:
# Initialize TRSTR-Paraphrase pipeline
pipe_trstr_para = HybridJournalPipeline(
    model_path=trstr_para_best, terms_csv_path="data/terms.csv"
)

# 1. Evaluate on Real Held-Out Test Set
test_trstr_para = evaluate_test_docbin(
    pipe_trstr_para, "data/training/test.spacy", run_label="trstr_paraphrase"
)

# 2. Evaluate on Unseen Benchmark
unseen_trstr_para = evaluate_unseen_generalization(
    pipe_trstr_para, run_label="trstr_paraphrase"
)

# Print TRSTR-Paraphrase Results
print("                    TRSTR-PARAPHRASE RESULTS")
print("=" * 75)
print(f"{'METRIC':<42} | {'TRSTR-Paraphrase':>18}")
print("-" * 75)
print(f"{'Held-Out Real Test Overall F1':<42} | {test_trstr_para['overall_f1']:>16}%")
print(f"{'Held-Out Real Test Overall Precision':<42} | {test_trstr_para['overall_precision']:>16}%")
print(f"{'Held-Out Real Test Overall Recall':<42} | {test_trstr_para['overall_recall']:>16}%")
print("-" * 75)
for lbl in ['IT_TERM', 'CLERICAL_TERM']:
    sl = test_trstr_para['labels'].get(lbl, {})
    print(f"  {lbl + ' F1':<40} | {sl.get('f1', 0):>16}%")
    print(f"  {lbl + ' Precision':<40} | {sl.get('precision', 0):>16}%")
    print(f"  {lbl + ' Recall':<40} | {sl.get('recall', 0):>16}%")
print("-" * 75)

b_trf_p = unseen_trstr_para['modes']['transformer_only']
print(f"{'Unseen Benchmark Transformer Recall':<42} | {b_trf_p.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer Precision':<42} | {b_trf_p.get('precision_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer F1':<42} | {b_trf_p.get('f1_pct', 0):>16}%")
b_hyb_p = unseen_trstr_para['modes']['hybrid']
print(f"{'Unseen Benchmark Hybrid Recall':<42} | {b_hyb_p.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Hybrid F1':<42} | {b_hyb_p.get('f1_pct', 0):>16}%")
print(f"{'Unseen Benchmark (Generalization Lift)':<42} | {unseen_trstr_para.get('generalization_lift', 'N/A'):>17}")
print("=" * 75)

[2026-10-01 05:59:07,290] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 05:59:07,291] INFO: Loading transformer model from 'models/ner_trf_trstr_paraphrase/model-best'...
[2026-10-01 05:59:10,386] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 05:59:19,219] INFO: Per-item results: data/eval_results/held_out_test_trstr_paraphrase.jsonl (187 items, 65 with errors)
[2026-10-01 05:59:19,323] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-10-01 05:59:19,324] INFO: Evaluating unseen benchmark: Mode 1/3 (Transformer-only)...
[2026-10-01 05:59:22,514] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_trstr_paraphrase.jsonl (85 items, 65 with errors)
[2026-10-01 05:59:22,515] INFO: Evaluating unseen benchmark: Mode 2/3 (EntityRuler-only)...
[2026-10-01 05:59:22,536] INFO: Per-item results: data/eval_results/unseen_benchmark_entity_ruler_only_trstr_paraphrase.jsonl (85 items, 66 with er

                    TRSTR-PARAPHRASE RESULTS
METRIC                                     |   TRSTR-Paraphrase
---------------------------------------------------------------------------
Held-Out Real Test Overall F1              |            70.06%
Held-Out Real Test Overall Precision       |            68.89%
Held-Out Real Test Overall Recall          |            71.26%
---------------------------------------------------------------------------
  IT_TERM F1                               |            70.78%
  IT_TERM Precision                        |            70.49%
  IT_TERM Recall                           |            71.07%
  CLERICAL_TERM F1                         |            68.47%
  CLERICAL_TERM Precision                  |            65.52%
  CLERICAL_TERM Recall                     |             71.7%
---------------------------------------------------------------------------
Unseen Benchmark Transformer Recall        |            76.85%
Unseen Benchmark Transformer Prec

---
---
# CONDITION 3: TRSTR via LLM Generation (Gemini)
---
## Phase 11: LLM-Direct Synthetic Data Generation

Generate novel sentence + entity pairs using Google Gemini (`gemini-3.5-flash-lite`) via `scripts/generate_llm_synthetic.py`.

Key design:
- LLM generates novel sentences (NOT paraphrases of existing ones)
- Grounded in annotation guidelines, terms.csv, and style examples from real data
- Entity offsets computed post-generation via exact string search (never by LLM)
- Quality filters: generic noun blocklist, benchmark term isolation, dedup guard

In [16]:
from scripts.generate_llm_synthetic import run_generation

LLM_SYNTHETIC_PATH = "data/synthetic_llm_generated.jsonl"

# Check if synthetic data already exists and is sufficient
existing_count = 0
if os.path.exists(LLM_SYNTHETIC_PATH):
    with open(LLM_SYNTHETIC_PATH, "r", encoding='utf-8') as f:
        existing_count = sum(1 for line in f if line.strip())

TARGET_SYNTHETIC = 300  # Target: 300 novel synthetic records
NUM_BATCHES = 30
BATCH_SIZE = 10

if existing_count >= TARGET_SYNTHETIC:
    print(f"Sufficient LLM-generated synthetic data already exists: {existing_count} records")
    print(f"Skipping generation. To regenerate, delete {LLM_SYNTHETIC_PATH} and re-run.")
else:
    print(f"Generating LLM synthetic data ({existing_count} existing, target {TARGET_SYNTHETIC})...")
    print(f"This calls the Gemini API and may take 5-10 minutes depending on rate limits.")
    llm_records = run_generation(
        num_batches=NUM_BATCHES,
        batch_size=BATCH_SIZE,
        output_path=LLM_SYNTHETIC_PATH,
        model="gemini-3.5-flash-lite",
    )
    print(f"\nLLM generation complete: {len(llm_records)} records saved.")

Sufficient LLM-generated synthetic data already exists: 300 records
Skipping generation. To regenerate, delete data/synthetic_llm_generated.jsonl and re-run.


---
## Phase 12: Build TRSTR-LLM Training Dataset & Train

Combine real training records with LLM-generated synthetic records (and optionally mined hard negatives), then fine-tune a fresh transformer.

In [17]:
from scripts.prepare_trstr_llm import build_trstr_llm_dataset

trstr_llm_result = build_trstr_llm_dataset(
    train_spacy_path="data/training/train.spacy",
    synthetic_jsonl_path=LLM_SYNTHETIC_PATH,
    mined_negatives_path="data/review/mined_hard_negatives.jsonl",
    dev_spacy_path="data/training/dev.spacy",
    test_spacy_path="data/training/test.spacy",
    terms_csv_path="data/terms.csv",
    output_jsonl_path="data/training_trstr_llm.jsonl",
    output_spacy_path="data/training/train_trstr_llm.spacy",
)

print(f"\nTRSTR-LLM Training Pool Summary:")
print(f"  Total records      : {trstr_llm_result['total_records']}")
print(f"  Real records       : {trstr_llm_result['real_count']}")
print(f"  Synthetic records  : {trstr_llm_result['synthetic_count']}")
print(f"  Mined negatives    : {trstr_llm_result['mined_count']}")
print(f"  Negative ratio     : {trstr_llm_result['negative_ratio_pct']}%")
print(f"  Output DocBin      : {trstr_llm_result['output_spacy']}")

[2026-10-01 05:59:26,334] INFO: Loaded 868 real training documents from data/training/train.spacy
[2026-10-01 05:59:26,343] INFO: Loaded 300 synthetic records from data/synthetic_llm_generated.jsonl
[2026-10-01 05:59:26,345] INFO: Loaded 144 raw mined hard negative entries from data/review/mined_hard_negatives.jsonl
[2026-10-01 05:59:26,404] INFO: Mined negatives curation audit: 144 raw -> 24 intra-file duplicates removed, 0 held-out test leaks blocked, 0 dev split leaks blocked, 14 accidental entity sentences filtered -> 106 clean hard negatives accepted.
[2026-10-01 05:59:26,405] INFO: Combined TRSTR-LLM pool: 1274 total records (868 real + 300 synthetic + 106 mined negatives)
[2026-10-01 05:59:26,413] INFO: Saved 1274 records to data/training_trstr_llm.jsonl
[2026-10-01 05:59:26,527] INFO: Saved combined DocBin to data/training/train_trstr_llm.spacy (1274 total docs, 0 synthetic entities dropped)
[2026-10-01 05:59:26,528] INFO: --- Real Training Split Diagnostics ---
[2026-10-01 05:


TRSTR-LLM Training Pool Summary:
  Total records      : 1274
  Real records       : 868
  Synthetic records  : 300
  Mined negatives    : 106
  Negative ratio     : 41.05%
  Output DocBin      : data/training/train_trstr_llm.spacy


In [18]:
# Free GPU memory and reload training module
import gc
import torch
import importlib
import scripts.training
importlib.reload(scripts.training)
from scripts.training import train_ner_trf

for var_name in ["pipe_trtr", "pipe_trstr_para", "pipe_trstr_llm", "para_model", "para_tokenizer"]:
    if var_name in globals():
        del globals()[var_name]
gc.collect()
if torch.cuda.is_available():
    torch.cuda.empty_cache()

# Train TRSTR-LLM model from scratch
trstr_llm_model_dir = "models/ner_trf_trstr_llm"
trstr_llm_best = os.path.join(trstr_llm_model_dir, "model-best")

# Clean previous model to train from scratch
if os.path.exists(trstr_llm_model_dir):
    shutil.rmtree(trstr_llm_model_dir)
    print(f"Cleaned previous TRSTR-LLM model at {trstr_llm_model_dir}")

print("Fine-tuning TRSTR-LLM model...")
train_ner_trf(
    output_dir=trstr_llm_model_dir,
    train_path="data/training/train_trstr_llm.spacy",
    dev_path="data/training/dev.spacy",
    config_path="config_trf.cfg",
    use_gpu=0,
    max_steps=2500,
    eval_frequency=50,
    patience=400
)
print(f"TRSTR-LLM model trained. Best checkpoint: {trstr_llm_best}")


[2026-10-01 05:59:29,209] INFO: Starting transformer training on GPU 0 for up to 2500 steps (eval_frequency=50, patience=400 steps, max_batch_items=2048)...
[2026-10-01 05:59:29,316] [INFO] Set up nlp object from config
[2026-10-01 05:59:29,316] INFO: Set up nlp object from config
[2026-10-01 05:59:29,328] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 05:59:29,328] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 05:59:29,333] [INFO] Created vocabulary
[2026-10-01 05:59:29,333] INFO: Created vocabulary
[2026-10-01 05:59:29,334] [INFO] Finished initializing nlp object
[2026-10-01 05:59:29,334] INFO: Finished initializing nlp object


Cleaned previous TRSTR-LLM model at models/ner_trf_trstr_llm
Fine-tuning TRSTR-LLM model...
ℹ Saving to output directory: models/ner_trf_trstr_llm
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 05:59:31,260] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 05:59:31,260] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0        2090.85    218.49    1.90    1.63    2.28    0.02
  2      50       73998.37  23782.39    0.00    0.00    0.00    0.00
  5     100       14242.50   6430.54   16.78   31.65   11.42    0.17
  7     150        6704.08   4929.76   32.16   35.75   29.22    0.32
 10     200       38971.56   8584.36   15.76   23.42   11.87    0.16
 12     250        9927.68   4855.21   44.69   43.35   46.12    0.45
 15     300        2902.22   2993.97   56.48   57.28   55.71    0.56
 17     350        1439.26   2185.75   60.58   59.13   62.10    0.61
 20     400        1142.78   1815.24   61.28   56.06   67.58    0.61
 22     450        1049.26   1801.85   66.81   64.02   69.86    0.67
 25     5

[2026-10-01 06:13:01,996] INFO: Training completed. Best model checkpoint: models/ner_trf_trstr_llm/model-best


✔ Saved pipeline to output directory
models/ner_trf_trstr_llm/model-last
TRSTR-LLM model trained. Best checkpoint: models/ner_trf_trstr_llm/model-best


---
## Phase 13: Evaluation (TRSTR-LLM)

Evaluate the TRSTR-LLM model on the same real evaluation partitions.

In [19]:
# Initialize TRSTR-LLM pipeline
pipe_trstr_llm = HybridJournalPipeline(
    model_path=trstr_llm_best, terms_csv_path="data/terms.csv"
)

# 1. Evaluate on Real Held-Out Test Set
test_trstr_llm = evaluate_test_docbin(
    pipe_trstr_llm, "data/training/test.spacy", run_label="trstr_llm"
)

# 2. Evaluate on Unseen Benchmark
unseen_trstr_llm = evaluate_unseen_generalization(
    pipe_trstr_llm, run_label="trstr_llm"
)

# Print TRSTR-LLM Results
print("                    TRSTR-LLM RESULTS")
print("=" * 75)
print(f"{'METRIC':<42} | {'TRSTR-LLM':>18}")
print("-" * 75)
print(f"{'Held-Out Real Test Overall F1':<42} | {test_trstr_llm['overall_f1']:>16}%")
print(f"{'Held-Out Real Test Overall Precision':<42} | {test_trstr_llm['overall_precision']:>16}%")
print(f"{'Held-Out Real Test Overall Recall':<42} | {test_trstr_llm['overall_recall']:>16}%")
print("-" * 75)
for lbl in ['IT_TERM', 'CLERICAL_TERM']:
    sl = test_trstr_llm['labels'].get(lbl, {})
    print(f"  {lbl + ' F1':<40} | {sl.get('f1', 0):>16}%")
    print(f"  {lbl + ' Precision':<40} | {sl.get('precision', 0):>16}%")
    print(f"  {lbl + ' Recall':<40} | {sl.get('recall', 0):>16}%")
print("-" * 75)

b_trf_l = unseen_trstr_llm['modes']['transformer_only']
print(f"{'Unseen Benchmark Transformer Recall':<42} | {b_trf_l.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer Precision':<42} | {b_trf_l.get('precision_pct', 0):>16}%")
print(f"{'Unseen Benchmark Transformer F1':<42} | {b_trf_l.get('f1_pct', 0):>16}%")
b_hyb_l = unseen_trstr_llm['modes']['hybrid']
print(f"{'Unseen Benchmark Hybrid Recall':<42} | {b_hyb_l.get('recall_pct', 0):>16}%")
print(f"{'Unseen Benchmark Hybrid F1':<42} | {b_hyb_l.get('f1_pct', 0):>16}%")
print(f"{'Unseen Benchmark (Generalization Lift)':<42} | {unseen_trstr_llm.get('generalization_lift', 'N/A'):>17}")
print("=" * 75)

[2026-10-01 06:13:02,028] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 06:13:02,029] INFO: Loading transformer model from 'models/ner_trf_trstr_llm/model-best'...
[2026-10-01 06:13:04,917] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 06:13:13,784] INFO: Per-item results: data/eval_results/held_out_test_trstr_llm.jsonl (187 items, 77 with errors)
[2026-10-01 06:13:13,888] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-10-01 06:13:13,888] INFO: Evaluating unseen benchmark: Mode 1/3 (Transformer-only)...
[2026-10-01 06:13:16,604] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_trstr_llm.jsonl (85 items, 51 with errors)
[2026-10-01 06:13:16,604] INFO: Evaluating unseen benchmark: Mode 2/3 (EntityRuler-only)...
[2026-10-01 06:13:16,626] INFO: Per-item results: data/eval_results/unseen_benchmark_entity_ruler_only_trstr_llm.jsonl (85 items, 66 with errors)
[2026-10-01 06:13:16,6

                    TRSTR-LLM RESULTS
METRIC                                     |          TRSTR-LLM
---------------------------------------------------------------------------
Held-Out Real Test Overall F1              |            64.25%
Held-Out Real Test Overall Precision       |             62.5%
Held-Out Real Test Overall Recall          |            66.09%
---------------------------------------------------------------------------
  IT_TERM F1                               |            66.95%
  IT_TERM Precision                        |             67.8%
  IT_TERM Recall                           |            66.12%
  CLERICAL_TERM F1                         |            58.82%
  CLERICAL_TERM Precision                  |            53.03%
  CLERICAL_TERM Recall                     |            66.04%
---------------------------------------------------------------------------
Unseen Benchmark Transformer Recall        |            72.22%
Unseen Benchmark Transformer Precision  

---
---
# COMPARATIVE ANALYSIS
---
## Phase 14: Side-by-Side Comparative Evaluation Table

Compare all 3 conditions (TRTR, TRSTR-Paraphrase, TRSTR-LLM) on the same held-out test set and unseen benchmark.

In [20]:
def fmt(v, suffix="%"):
    """Format a numeric value for table display."""
    if isinstance(v, (int, float)):
        return f"{v}{suffix}"
    return str(v)

print("\n" + "=" * 100)
print("          SIDE-BY-SIDE COMPARATIVE EVALUATION — ALL 3 CONDITIONS")
print("=" * 100)
print(f"{'EVALUATION METRIC':<45} | {'TRTR':>14} | {'TRSTR-Para':>14} | {'TRSTR-LLM':>14}")
print("-" * 100)

# Held-out real test set
print(f"{'Held-Out Test Overall F1':<45} | {fmt(test_trtr['overall_f1']):>14} | {fmt(test_trstr_para['overall_f1']):>14} | {fmt(test_trstr_llm['overall_f1']):>14}")
print(f"{'Held-Out Test Overall Precision':<45} | {fmt(test_trtr['overall_precision']):>14} | {fmt(test_trstr_para['overall_precision']):>14} | {fmt(test_trstr_llm['overall_precision']):>14}")
print(f"{'Held-Out Test Overall Recall':<45} | {fmt(test_trtr['overall_recall']):>14} | {fmt(test_trstr_para['overall_recall']):>14} | {fmt(test_trstr_llm['overall_recall']):>14}")
print("-" * 100)

# Per-label
for lbl in ['IT_TERM', 'CLERICAL_TERM']:
    for metric_key, metric_name in [('f1', 'F1'), ('precision', 'Precision'), ('recall', 'Recall')]:
        v_trtr = test_trtr['labels'].get(lbl, {}).get(metric_key, 0)
        v_para = test_trstr_para['labels'].get(lbl, {}).get(metric_key, 0)
        v_llm = test_trstr_llm['labels'].get(lbl, {}).get(metric_key, 0)
        print(f"  {lbl + ' ' + metric_name:<43} | {fmt(v_trtr):>14} | {fmt(v_para):>14} | {fmt(v_llm):>14}")
print("-" * 100)

# Unseen benchmark - Transformer only
trf_trtr = unseen_trtr['modes']['transformer_only']
trf_para = unseen_trstr_para['modes']['transformer_only']
trf_llm = unseen_trstr_llm['modes']['transformer_only']

print(f"{'Unseen Benchmark TRF Recall':<45} | {fmt(trf_trtr.get('recall_pct', 0)):>14} | {fmt(trf_para.get('recall_pct', 0)):>14} | {fmt(trf_llm.get('recall_pct', 0)):>14}")
print(f"{'Unseen Benchmark TRF Precision':<45} | {fmt(trf_trtr.get('precision_pct', 0)):>14} | {fmt(trf_para.get('precision_pct', 0)):>14} | {fmt(trf_llm.get('precision_pct', 0)):>14}")
print(f"{'Unseen Benchmark TRF F1':<45} | {fmt(trf_trtr.get('f1_pct', 0)):>14} | {fmt(trf_para.get('f1_pct', 0)):>14} | {fmt(trf_llm.get('f1_pct', 0)):>14}")
print("-" * 100)

# Unseen benchmark - Hybrid
hyb_trtr = unseen_trtr['modes']['hybrid']
hyb_para = unseen_trstr_para['modes']['hybrid']
hyb_llm = unseen_trstr_llm['modes']['hybrid']

print(f"{'Unseen Benchmark Hybrid Recall':<45} | {fmt(hyb_trtr.get('recall_pct', 0)):>14} | {fmt(hyb_para.get('recall_pct', 0)):>14} | {fmt(hyb_llm.get('recall_pct', 0)):>14}")
print(f"{'Unseen Benchmark Hybrid F1':<45} | {fmt(hyb_trtr.get('f1_pct', 0)):>14} | {fmt(hyb_para.get('f1_pct', 0)):>14} | {fmt(hyb_llm.get('f1_pct', 0)):>14}")
print(f"{'Generalization Lift':<45} | {unseen_trtr.get('generalization_lift', 'N/A'):>14} | {unseen_trstr_para.get('generalization_lift', 'N/A'):>14} | {unseen_trstr_llm.get('generalization_lift', 'N/A'):>14}")

print("=" * 100)

# Compute deltas vs TRTR baseline
print("\n" + "-" * 100)
print("  DELTA vs. TRTR BASELINE")
print("-" * 100)
for name, test_res, unseen_res in [
    ("TRSTR-Paraphrase", test_trstr_para, unseen_trstr_para),
    ("TRSTR-LLM", test_trstr_llm, unseen_trstr_llm),
]:
    delta_f1 = round(test_res['overall_f1'] - test_trtr['overall_f1'], 2)
    delta_p = round(test_res['overall_precision'] - test_trtr['overall_precision'], 2)
    delta_r = round(test_res['overall_recall'] - test_trtr['overall_recall'], 2)
    
    trf_mode = unseen_res['modes']['transformer_only']
    trf_base = unseen_trtr['modes']['transformer_only']
    delta_unseen_r = round(trf_mode.get('recall_pct', 0) - trf_base.get('recall_pct', 0), 2)
    delta_unseen_f1 = round(trf_mode.get('f1_pct', 0) - trf_base.get('f1_pct', 0), 2)
    
    sign = lambda x: f"+{x}" if x > 0 else str(x)
    print(f"  {name:<20}: Test F1 {sign(delta_f1)}% | Test P {sign(delta_p)}% | "
          f"Test R {sign(delta_r)}% | Unseen TRF Recall {sign(delta_unseen_r)}% | "
          f"Unseen TRF F1 {sign(delta_unseen_f1)}%")

print("=" * 100)


          SIDE-BY-SIDE COMPARATIVE EVALUATION — ALL 3 CONDITIONS
EVALUATION METRIC                             |           TRTR |     TRSTR-Para |      TRSTR-LLM
----------------------------------------------------------------------------------------------------
Held-Out Test Overall F1                      |         67.63% |         70.06% |         64.25%
Held-Out Test Overall Precision               |         68.02% |         68.89% |          62.5%
Held-Out Test Overall Recall                  |         67.24% |         71.26% |         66.09%
----------------------------------------------------------------------------------------------------
  IT_TERM F1                                  |          67.8% |         70.78% |         66.95%
  IT_TERM Precision                           |         69.57% |         70.49% |          67.8%
  IT_TERM Recall                              |         66.12% |         71.07% |         66.12%
  CLERICAL_TERM F1                            |      

---
## Phase 15: Save Full Evaluation Reports

Export comprehensive JSON reports for each condition.

In [21]:
# Save comprehensive comparison report
comparison_report = {
    "conditions": {
        "TRTR": {
            "description": "Train Real, Test Real — baseline with authentic data only",
            "training_records": len(train_records),
            "held_out_test": test_trtr,
            "unseen_benchmark": {
                "transformer_only": unseen_trtr['modes']['transformer_only'],
                "hybrid": unseen_trtr['modes']['hybrid'],
                "generalization_lift": unseen_trtr.get('generalization_lift'),
            },
        },
        "TRSTR_Paraphrase": {
            "description": "Train Real + T5 Paraphrases, Test Real",
            "training_records": len(combined_para_records),
            "real_records": len(train_records),
            "synthetic_records": len(paraphrase_records),
            "held_out_test": test_trstr_para,
            "unseen_benchmark": {
                "transformer_only": unseen_trstr_para['modes']['transformer_only'],
                "hybrid": unseen_trstr_para['modes']['hybrid'],
                "generalization_lift": unseen_trstr_para.get('generalization_lift'),
            },
        },
        "TRSTR_LLM": {
            "description": "Train Real + LLM-Direct Gemini Generation, Test Real",
            "training_records": trstr_llm_result['total_records'],
            "real_records": trstr_llm_result['real_count'],
            "synthetic_records": trstr_llm_result['synthetic_count'],
            "mined_negatives": trstr_llm_result['mined_count'],
            "held_out_test": test_trstr_llm,
            "unseen_benchmark": {
                "transformer_only": unseen_trstr_llm['modes']['transformer_only'],
                "hybrid": unseen_trstr_llm['modes']['hybrid'],
                "generalization_lift": unseen_trstr_llm.get('generalization_lift'),
            },
        },
    },
    "evaluation_protocol": {
        "training_hyperparameters": {
            "max_steps": 2500,
            "eval_frequency": 50,
            "patience": 400,
            "backbone": "roberta-base",
        },
        "test_set": "data/training/test.spacy (100% authentic, zero synthetic)",
        "dev_set": "data/training/dev.spacy (100% authentic, zero synthetic)",
        "unseen_benchmark": "data/test/unseen_benchmark.jsonl (65 unseen terms)",
    },
}

report_path = "data/evaluation_report_3way_comparison.json"
with open(report_path, "w", encoding="utf-8") as f:
    json.dump(comparison_report, f, indent=2, ensure_ascii=False)

print(f"Full 3-way comparison report saved to {report_path}")

Full 3-way comparison report saved to data/evaluation_report_3way_comparison.json


---
## Phase 16: Findings & Thesis Methodological Summary

### Comparative Conclusions:

| Condition | Description | Training Pool |
| :--- | :--- | :--- |
| **TRTR** | Baseline — authentic data only | Real training split only |
| **TRSTR-Paraphrase** | T5 paraphrase augmentation of real sentences | Real + T5 paraphrases |
| **TRSTR-LLM** | LLM-direct novel sentence generation (Gemini) | Real + Gemini synthetic |

All 3 conditions were evaluated under **identical hyperparameters** (`max_steps=2500`, `patience=400`, `eval_frequency=50`) on the **exact same real evaluation partitions** (dev/test splits, unseen benchmark).

> **Strict Evaluation Isolation**: Synthetic data exists **strictly on the training side**. All validation and test sets consist solely of authentic records. Synthetic data never touches the evaluation side.

> [!WARNING]
> **Limitations Disclaimer**: Any synthetic augmentation is introduced to compensate for the limited volume of the authentic annotated corpus. TRTR remains the reference point for real-data-only performance. Augmented results should not be misconstrued as evidence that the system was trained exclusively on authentic data.

---
---
# CROSS-VALIDATION
---
## Phase 17: 5-Fold Cross-Validation (TRTR vs. TRSTR-Paraphrase vs. TRSTR-LLM)

To establish statistical significance beyond a single train/dev/test split, this phase runs **Stratified K-Fold Cross-Validation** over all 1,241 authentic records in `data/data.jsonl`.

### Rigorous Protocol:
1. **Pure Real Validation**: In each fold, 20% of authentic records serve as the held-out validation fold. Synthetic data **never** enters the validation fold.
2. **Three Canonical Conditions**:
   - **`TRTR`**: Real training fold only.
   - **`TRSTR-Paraphrase`**: Real training fold + T5 synthetic paraphrases (strictly filtered against validation sentences).
   - **`TRSTR-LLM`**: Real training fold + Gemini LLM synthetic data + curated mined hard negatives (strictly filtered against validation sentences).
3. **Statistical Aggregation**: Reports **Mean ± Standard Deviation** across all folds for Strict F1, Precision, Recall, and Unseen Benchmark Generalization.
4. **Resume-Safe Caching**: Fold checkpoints and per-fold metrics are saved under `data/cv/results/` so the run can be resumed seamlessly.


In [22]:
# Free GPU memory before cross-validation
import gc
import torch
import importlib
import scripts.cross_validation
importlib.reload(scripts.cross_validation)
from scripts.cross_validation import run_3way_cross_validation

for var_name in ["pipe_trtr", "pipe_trstr_para", "pipe_trstr_llm", "pipeline"]:
    if var_name in globals():
        del globals()[var_name]
gc.collect()
if torch.cuda.is_available():
    torch.cuda.empty_cache()

# Execute 3-way K-Fold Cross-Validation
# Note: Per-fold results are cached in data/cv/results/ so execution is resume-safe.
cv_results = run_3way_cross_validation(
    data_path="data/data.jsonl",
    terms_csv_path="data/terms.csv",
    unseen_benchmark_path="data/test/unseen_benchmark.jsonl",
    n_splits=5,
    conditions=["trtr", "trstr_paraphrase", "trstr_llm"],
    max_steps=1500,
    eval_frequency=50,
    patience=250,
    force_train=False
)


[2026-10-01 06:13:21,311] INFO: CV folds already exist at data/cv. Loading metadata.
[2026-10-01 06:13:21,313] INFO: Loaded 85 unseen benchmark samples from data/test/unseen_benchmark.jsonl
[2026-10-01 06:13:21,313] INFO: 
STARTING CV EVALUATION: TRTR
[2026-10-01 06:13:21,314] INFO: Loaded cached run results from data/cv/results/trtr_f0.json
[2026-10-01 06:13:21,315] INFO: Loaded cached run results from data/cv/results/trtr_f1.json
[2026-10-01 06:13:21,316] INFO: Loaded cached run results from data/cv/results/trtr_f2.json
[2026-10-01 06:13:21,316] INFO: Loaded cached run results from data/cv/results/trtr_f3.json
[2026-10-01 06:13:21,317] INFO: Loaded cached run results from data/cv/results/trtr_f4.json
[2026-10-01 06:13:21,317] INFO: 
STARTING CV EVALUATION: TRSTR_PARAPHRASE
[2026-10-01 06:13:21,318] INFO: Loaded cached run results from data/cv/results/trstr_paraphrase_f0.json
[2026-10-01 06:13:21,319] INFO: Loaded cached run results from data/cv/results/trstr_paraphrase_f1.json
[2026-

ℹ Saving to output directory: models/cv/trstr_paraphrase_f4
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 06:13:40,908] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 06:13:40,908] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0         340.40    121.08    2.04    1.18    7.55    0.02
  1      50       96726.11  44764.87    0.00    0.00    0.00    0.00
  3     100        6712.69   6126.26    0.00    0.00    0.00    0.00
  5     150        3369.57   4755.92   21.39   20.27   22.64    0.21
  7     200        3451.96   4239.15   34.34   44.58   27.92    0.34
  9     250        2718.24   3567.98   42.42   40.20   44.91    0.42
 11     300        2520.08   3219.11   51.81   49.83   53.96    0.52
 13     350        1791.70   2481.69   63.24   66.39   60.38    0.63
 15     400        1657.14   2151.10   68.56   68.82   68.30    0.69
 17     450        1034.66   1673.32   66.53   72.44   61.51    0.67
 19     5

[2026-10-01 06:24:19,885] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 06:24:19,886] INFO: Loading transformer model from 'models/cv/trstr_paraphrase_f4/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_paraphrase_f4/model-last


[2026-10-01 06:24:22,673] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 06:24:34,124] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_062434.jsonl (85 items, 62 with errors)
[2026-10-01 06:24:37,195] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_062437.jsonl (85 items, 62 with errors)
[2026-10-01 06:24:38,749] INFO: 
STARTING CV EVALUATION: TRSTR_LLM
[2026-10-01 06:24:38,943] INFO: Training TRSTR_LLM (Fold 0) for up to 1500 steps...
[2026-10-01 06:24:39,051] [INFO] Set up nlp object from config
[2026-10-01 06:24:39,051] INFO: Set up nlp object from config
[2026-10-01 06:24:39,063] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 06:24:39,063] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 06:24:39,068] [INFO] Created vocabulary
[2026-10-01 06:24:39,068] INFO: Created vocabulary
[2026-10-01 06:24:39,068] [INFO] Finished initializing nlp object
[2026-10-01 06:24:39,068] INFO: Finished

ℹ Saving to output directory: models/cv/trstr_llm_f0
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 06:24:40,648] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 06:24:40,648] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0         873.10    318.84    2.66    1.54    9.45    0.03
  2      50      103001.11  44263.18    0.00    0.00    0.00    0.00
  4     100       14346.71   7115.28    0.00    0.00    0.00    0.00
  6     150       22473.83   6178.65   24.70   27.85   22.18    0.25
  9     200       43266.91   8098.92   16.37   27.59   11.64    0.16
 11     250        2862.20   3455.46   48.21   53.30   44.00    0.48
 13     300        2134.83   2594.67   52.73   70.30   42.18    0.53
 15     350        1930.00   2339.07   39.42   39.56   39.27    0.39
 18     400        1181.42   1668.25   60.84   60.95   60.73    0.61
 20     450        1127.24   1352.23   62.61   61.62   63.64    0.63
 22     5

[2026-10-01 06:31:24,137] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 06:31:24,138] INFO: Loading transformer model from 'models/cv/trstr_llm_f0/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_llm_f0/model-last


[2026-10-01 06:31:26,946] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 06:31:38,295] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_063138.jsonl (85 items, 58 with errors)
[2026-10-01 06:31:41,373] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_063141.jsonl (85 items, 58 with errors)
[2026-10-01 06:31:42,632] INFO: Training TRSTR_LLM (Fold 1) for up to 1500 steps...
[2026-10-01 06:31:42,743] [INFO] Set up nlp object from config
[2026-10-01 06:31:42,743] INFO: Set up nlp object from config
[2026-10-01 06:31:42,755] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 06:31:42,755] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 06:31:42,759] [INFO] Created vocabulary
[2026-10-01 06:31:42,759] INFO: Created vocabulary
[2026-10-01 06:31:42,760] [INFO] Finished initializing nlp object
[2026-10-01 06:31:42,760] INFO: Finished initializing nlp object


ℹ Saving to output directory: models/cv/trstr_llm_f1
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 06:31:44,452] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 06:31:44,452] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0        1601.34    370.33    2.27    1.29    9.20    0.02
  2      50       93792.10  45173.95    0.00    0.00    0.00    0.00
  4     100        4361.85   6450.35    0.00    0.00    0.00    0.00
  6     150        4028.49   5061.04   20.86   44.74   13.60    0.21
  8     200        3646.91   4334.46   40.16   39.53   40.80    0.40
 11     250        4632.43   4549.16   52.61   57.62   48.40    0.53
 13     300        2095.55   2627.38   53.43   68.99   43.60    0.53
 15     350        1562.83   2150.70   57.32   61.09   54.00    0.57
 17     400        1555.73   1997.56   61.47   66.98   56.80    0.61
 20     450        1030.90   1576.84   61.86   58.84   65.20    0.62
 22     5

[2026-10-01 06:41:40,348] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 06:41:40,348] INFO: Loading transformer model from 'models/cv/trstr_llm_f1/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_llm_f1/model-last


[2026-10-01 06:41:42,820] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 06:41:53,073] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_064153.jsonl (85 items, 52 with errors)
[2026-10-01 06:41:55,642] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_064155.jsonl (85 items, 53 with errors)
[2026-10-01 06:41:57,295] INFO: Training TRSTR_LLM (Fold 2) for up to 1500 steps...
[2026-10-01 06:41:57,406] [INFO] Set up nlp object from config
[2026-10-01 06:41:57,406] INFO: Set up nlp object from config
[2026-10-01 06:41:57,419] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 06:41:57,419] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 06:41:57,423] [INFO] Created vocabulary
[2026-10-01 06:41:57,423] INFO: Created vocabulary
[2026-10-01 06:41:57,424] [INFO] Finished initializing nlp object
[2026-10-01 06:41:57,424] INFO: Finished initializing nlp object


ℹ Saving to output directory: models/cv/trstr_llm_f2
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 06:41:59,083] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 06:41:59,083] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0         779.12    286.72    2.29    1.32    8.46    0.02
  2      50       96890.27  43248.22    0.00    0.00    0.00    0.00
  4     100        4474.90   6350.66    0.00    0.00    0.00    0.00
  6     150       14639.68   7090.80   10.58   21.84    6.99    0.11
  9     200        3442.67   4620.73   28.19   35.16   23.53    0.28
 11     250        4436.51   4825.01   30.50   25.06   38.97    0.31
 13     300        2901.25   3828.97   35.29   45.88   28.68    0.35
 15     350        2341.95   3067.85   59.93   57.88   62.13    0.60
 18     400        1575.17   2367.87   59.74   59.21   60.29    0.60
 20     450        1324.42   2092.17   61.45   62.26   60.66    0.61
 22     5

[2026-10-01 06:52:36,343] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 06:52:36,343] INFO: Loading transformer model from 'models/cv/trstr_llm_f2/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_llm_f2/model-last


[2026-10-01 06:52:39,115] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 06:52:49,621] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_065249.jsonl (85 items, 48 with errors)
[2026-10-01 06:52:52,229] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_065252.jsonl (85 items, 49 with errors)
[2026-10-01 06:52:53,921] INFO: Training TRSTR_LLM (Fold 3) for up to 1500 steps...
[2026-10-01 06:52:54,017] [INFO] Set up nlp object from config
[2026-10-01 06:52:54,017] INFO: Set up nlp object from config
[2026-10-01 06:52:54,028] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 06:52:54,028] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 06:52:54,032] [INFO] Created vocabulary
[2026-10-01 06:52:54,032] INFO: Created vocabulary
[2026-10-01 06:52:54,033] [INFO] Finished initializing nlp object
[2026-10-01 06:52:54,033] INFO: Finished initializing nlp object


ℹ Saving to output directory: models/cv/trstr_llm_f3
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 06:52:55,815] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 06:52:55,815] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0        1741.34    456.86    3.04    1.76   11.07    0.03
  2      50       95156.21  44634.54    0.00    0.00    0.00    0.00
  4     100        7717.08   6897.59    0.00    0.00    0.00    0.00
  6     150       21657.55   6284.51    9.52   13.43    7.38    0.10
  9     200        4019.13   4910.54   14.72   40.00    9.02    0.15
 11     250        3068.29   4053.24   21.45   46.58   13.93    0.21
 13     300        2672.79   3627.95   44.09   38.70   51.23    0.44
 15     350        3025.97   3430.29   49.06   49.79   48.36    0.49
 18     400        2038.29   2702.53   50.57   57.59   45.08    0.51
 20     450        1292.46   1981.78   57.89   57.20   58.61    0.58
 22     5

[2026-10-01 07:03:39,877] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 07:03:39,878] INFO: Loading transformer model from 'models/cv/trstr_llm_f3/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_llm_f3/model-last


[2026-10-01 07:03:42,380] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 07:03:52,743] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_070352.jsonl (85 items, 52 with errors)
[2026-10-01 07:03:55,440] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_070355.jsonl (85 items, 53 with errors)
[2026-10-01 07:03:57,173] INFO: Training TRSTR_LLM (Fold 4) for up to 1500 steps...
[2026-10-01 07:03:57,271] [INFO] Set up nlp object from config
[2026-10-01 07:03:57,271] INFO: Set up nlp object from config
[2026-10-01 07:03:57,282] [INFO] Pipeline: ['transformer', 'ner']
[2026-10-01 07:03:57,282] INFO: Pipeline: ['transformer', 'ner']
[2026-10-01 07:03:57,286] [INFO] Created vocabulary
[2026-10-01 07:03:57,286] INFO: Created vocabulary
[2026-10-01 07:03:57,287] [INFO] Finished initializing nlp object
[2026-10-01 07:03:57,287] INFO: Finished initializing nlp object


ℹ Saving to output directory: models/cv/trstr_llm_f4
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[2026-10-01 07:03:58,806] [INFO] Initialized pipeline components: ['transformer', 'ner']
[2026-10-01 07:03:58,806] INFO: Initialized pipeline components: ['transformer', 'ner']


✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.0
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
  0       0         247.23    412.78    3.27    1.74   27.55    0.03
  2      50       14784.45  25265.84    0.00    0.00    0.00    0.00
  4     100        2758.90   5267.45   21.91   42.86   14.72    0.22
  6     150        5213.50   4656.52   44.62   46.34   43.02    0.45
  9     200        2601.51   3455.51   40.00   59.26   30.19    0.40
 11     250        1838.14   2787.30   63.90   61.25   66.79    0.64
 13     300        1482.90   2328.52   60.85   59.85   61.89    0.61
 15     350        1243.32   2017.28   64.62   61.94   67.55    0.65
 18     400        1104.60   1857.94   62.90   62.78   63.02    0.63
 20     450         950.49   1671.66   60.20   64.78   56.23    0.60
 22     5

[2026-10-01 07:12:30,153] INFO: Loaded 343 unique terms from data/terms.csv
[2026-10-01 07:12:30,153] INFO: Loading transformer model from 'models/cv/trstr_llm_f4/model-best'...


✔ Saved pipeline to output directory
models/cv/trstr_llm_f4/model-last


[2026-10-01 07:12:32,612] INFO: Configured standalone Dictionary matcher with 343 patterns.
[2026-10-01 07:12:43,381] INFO: Per-item results: data/eval_results/unseen_benchmark_transformer_only_20261001_071243.jsonl (85 items, 52 with errors)
[2026-10-01 07:12:46,033] INFO: Per-item results: data/eval_results/unseen_benchmark_hybrid_20261001_071246.jsonl (85 items, 53 with errors)
[2026-10-01 07:12:47,287] INFO: Consolidated CV summary saved to: data/cv/cv_3way_results.json



      K-FOLD CROSS-VALIDATION SUMMARY (MEAN ± STD DEV)
METRIC                               | TRTR (Real-Only) | TRSTR-Paraphrase |        TRSTR-LLM
------------------------------------------------------------------------------------------
Held-Out Validation Overall F1       |  70.00 ± 3.72% |  74.97 ± 2.91% |  69.05 ± 3.24%
Held-Out Validation Overall Precision |  66.41 ± 3.70% |  72.26 ± 2.67% |  65.81 ± 3.60%
Held-Out Validation Overall Recall   |  74.08 ± 4.36% |  77.94 ± 3.53% |  72.65 ± 3.20%
  IT_TERM F1                         |  70.73 ± 3.50% |  74.84 ± 2.35% |  70.75 ± 3.10%
  CLERICAL_TERM F1                   |  68.51 ± 5.14% |  75.21 ± 4.61% |  65.69 ± 4.03%
Unseen Benchmark TRF Recall          |  76.00 ± 3.59% |  73.94 ± 5.33% |  76.11 ± 1.80%
Unseen Benchmark TRF F1              |  49.41 ± 1.93% |  58.45 ± 9.51% |  74.66 ± 0.71%
